universe/Condition.cpp

```diff
@@ -9955,6 +9955,203 @@ unsigned int Not::GetCheckSum() const {
     TraceLogger() << "GetCheckSum(Not): retval: " << retval;
     return retval;
 }
+///////////////////////////////////////////////////////////
+// TopmostMatches
+///////////////////////////////////////////////////////////
+TopmostMatches::TopmostMatches(std::vector<std::unique_ptr<ConditionBase>>&& operands) :
+    ConditionBase(),
+    m_operands(std::move(operands))
+{}
+
+bool TopmostMatches::operator==(const ConditionBase& rhs) const {
+    if (this == &rhs)
+        return true;
+    if (typeid(*this) != typeid(rhs))
+        return false;
+
+    const TopmostMatches& rhs_ = static_cast<const TopmostMatches&>(rhs);
+
+    if (m_operands.size() != rhs_.m_operands.size())
+        return false;
+    for (unsigned int i = 0; i < m_operands.size(); ++i) {
+        CHECK_COND_VREF_MEMBER(m_operands.at(i))
+    }
+
+    return true;
+}
+
+void TopmostMatches::Eval(const ScriptingContext& parent_context, ObjectSet& matches,
+               ObjectSet& non_matches, SearchDomain search_domain/* = NON_MATCHES*/) const
+{
+    std::shared_ptr<const UniverseObject> no_object;
+    ScriptingContext local_context(parent_context, no_object);
+
+    if (m_operands.empty()) {
+        ErrorLogger() << "TopmostMatches::Eval given no operands!";
+        return;
+    }
+    for (auto& operand : m_operands) {
+        if (!operand) {
+            ErrorLogger() << "TopmostMatches::Eval given null operand!";
+            return;
+        }
+    }
+
+    // probably i need to figure out which condition matches first
+    // by searching both domains
+    // ... or actually i can propagate the search domain and look for the first match, probably i can
+    // Expressions like  'NOT TopmostMatches [ A B C ]' matches all candidates which do not match the topmost matching condition
+    // which means the same as TopmostMatches [ (Not A) (Not B) (Not C) ]
```

Comment:
If the test is `Not TopmostMatches [All Ship]`, it should match nothing.

If the test is `TopmostMatches [Not All Not Ship]`, it should match `{planet 2, building 3}`.

Those aren't the same...

Reply:
Yeah you are right. I think NOT TopmostMatches [ A B C ] in FOCS terms should be the same as
```
TopmostMatches [ 
    And [
        Number low = 1 condition = A
        Not A
    ]
    And [
        Number low = 1 condition = B
        Not B
    ]
    And [
        Number low = 1 condition = C
        Not C
    ]
```
In the description, which is usually used to describe a mismatch it should probably read like:
```{ <<Not A>> iff there is at least one opposite match, else <<Not B>> iff there is at least one opposite match, else <<Not A>> iff there is at least one opposite match }```

or something like:
```matches the first conditions which has matching and non-matching candidates: { <<Not A>>, else <<Not B>>, else <<Not C>> }```

Reply:
Keep in mind that `Number` tests its subcondition on all objects in the universe, not (just) those in the candidate `matches` and `non_matches` sets.

https://github.com/freeorion/freeorion/blob/master/universe/Condition.cpp#L415

Reply:
So that woiuld mean rather ```Number low = 1 condition = And [ Object id = RootCandidate.ID  A ]```?

Hm. Not sure RootCandidate is available though.

Reply:
Would Filtering for Object I'd  = LocalCandidate.ID inside the Number condition suffice?

Reply:
Would Filtering for Object I'd  = LocalCandidate.ID inside the Number condition suffice?

Reply:
Would Filtering for Object I'd  = LocalCandidate.ID inside the Number condition suffice?

Reply:
I don't think that makes sense, and it's probably not supported. `LocalCandidate` within a subcondition of `Number` would be the objects being matched by that condition, and has nothing to do with objects being matched by `Number` itself. `Number` doesn't even pass a `LocalCandidate` to its subcondition. In some cases you might be able to do something with `RootCandidate`, but I don't advise it.

Reply:
No, i think the implementaton for `matches` and `non_matches` will be quite similar in the end and not depend on other conditions.

But i think it would be helpful for documentation purposes to express it in FOCS terms.

So I probably would need a hypethetical `OuterCandidate` which allows to pump the matches of the surrounding (so outta Number in this case) into the inner condition
```TopmostMatches [ 
    And [
        Number low = 1 condition = And [ 
            Object id = OuterCandidate.ID
            A
        ]
        Not A
    ]
    ...
]```